### PR TITLE
Deprecate `Transaction`'s `signature_hash`/`encode_signing_data_to` methods

### PR DESCRIPTION
**Pull request by tcharding**
When we introduced the `SighashCache` we put encoding and sighash code
for segwit (v0 and taproo) in the `sighash` module. We also added
methods for legacy inputs that wrapped calls to encode/sighash methods
in the `transaction` module. This is confusing for a couple of reasons

- When devs read `Transaction` there are two methods that apparently
enable encodeing tx data and calculating the sighash but there is no
indication that these methods are only for legacy inputs.

- Once devs work out that segwit inputs have to be handled by methods on
the `SighashCache` it is not obvious why the methods on `Transaction`
exist at all.

Move the legacy encode/sighash code over to the `sighash` module and
deprecate the old methods. (Includes moving unit tests.)

Patch 1 and 2 are preparation.

(I did this work because I got lost in `Transaction::signature_hash` when messing with `TxOut::Amount` and trying to understand the signing algorithm vs bip143.

sanket1729 is there some other reason that I'm missing why we left the legacy code in `transaction`?

### CODE REVIEW

File: src/util/sighash.rs [hunk 1]
```diff
@@ -639,11 +640,31 @@ impl<R: Deref<Target = Transaction>> SighashCache<R> {
...
+    /// # Warning
+    ///
+    /// - Does NOT attempt to support OP_CODESEPARATOR. In general this would require evaluating
+    /// `script_pubkey` to determine which separators get evaluated and which don't, which we don't
+    /// have the information to determine.
+    /// - Does NOT handle the sighash single bug (see "Return type" section)
+    ///
+    /// # Returns
+    ///
+    /// This function can handle the SIGHASH_SINGLE bug internally, so it returns [`EncodeSigningDataResult`]
```
Comment (apoelstra) [inline src/util/sighash.rs:663]: In 8cbc99fed84a48e02e1fd61d2ae1e630e82fab50:

`can handle` should be `can't handle`

**Review by apoelstra — DISMISSED** [review]
ACK 8cbc99fed84a48e02e1fd61d2ae1e630e82fab50

File: src/util/sighash.rs (hunk 1 above)
Comment (tcharding) [reply src/util/sighash.rs:663]: lolz, I think I edited that from "can't" to "cannot" then undid the edit - incorrectly. Thanks for picking it up.

**Comment by tcharding** [comment]
Changes in force push: Just the rustdoc fix suggested above.

**Review by apoelstra — APPROVED** [review]
ACK 03413fc4c274ae795847e72db8c1f21d602d29ca

**Comment by tcharding** [comment]
Rebase only, no other changes.

**Review by apoelstra — APPROVED** [review]
ACK 3113ebc879f51d0537b97de06e895b741aa64a03

**Comment by tcharding** [comment]
Rebase to pick up `serde` pinning change, no other changes.

**Review by sanket1729 — APPROVED** [review]
ACK c062e4ff809be66508979eb6aa4db1adfb4c87e8

**Review by apoelstra — APPROVED** [review]
ACK c062e4ff809be66508979eb6aa4db1adfb4c87e8

**Pull request merged**